### PR TITLE
2808 - converted js to es6 classes/modules, added arrow functions, var to const/let, added imports/exports, return early

### PR DESCRIPTION
**Pull request by eleanor-heath**
adaptlearning/adapt_framework#2808

### CODE REVIEW

**Review by cahirodoherty-learningpool — APPROVED** [review]
👀

File: js/adapt-youtube.js [hunk 1]
```diff
@@ -1,208 +1,206 @@
...
+      Adapt.youTubeIframeAPIReady = true;
+      Adapt.trigger('youTubeIframeAPIReady');
+    };
+    $.getScript('//www.youtube.com/iframe_api');
+  }
+
+  preRender() {
+    this.listenTo(Adapt, {
+      'device:resize': this.setIFrameSize,
+      'device:changed': this.setIFrameSize,
```
Comment (tomgreenfield) [inline js/adapt-youtube.js]:
```suggestion
      'device:resize device:changed': this.setIFrameSize,
```

File: js/adapt-youtube.js [hunk 2]
```diff
@@ -1,208 +1,206 @@
...
+  setIFrameSize() {
+    const $iframe = this.$('iframe');
+    const widgetWidth = this.$('.component__widget').width();
+
+    $iframe.width(widgetWidth);
+
+    // default aspect ratio to 16:9 if not specified
+    const aspectRatio = this.model.get('_media')._aspectRatio 
+      ? parseFloat(this.model.get('_media')._aspectRatio)
+      : 1.778;
```
Comment (tomgreenfield) [inline js/adapt-youtube.js]:
```suggestion
    const aspectRatio = parseFloat(this.model.get('_media')._aspectRatio) || 1.778;
```

File: js/adapt-youtube.js [hunk 3]
```diff
@@ -1,208 +1,206 @@
...
-    postRender: function() {
-      //FOR HTML/HBS Paramenters: https://developers.google.com/youtube/player_parameters
-      if (Adapt.youTubeIframeAPIReady === true) {
-        this.onYouTubeIframeAPIReady();
-      } else {
-        Adapt.once('youTubeIframeAPIReady', this.onYouTubeIframeAPIReady, this);
+  onYouTubeIframeAPIReady() {
+    this.player = new YT.Player(this.$('iframe').get(0), {
+      events: {
+        'onStateChange': this.onPlayerStateChange,
```
Comment (tomgreenfield) [inline js/adapt-youtube.js]: ESLint: remove quotes.

File: js/adapt-youtube.js [hunk 4]
```diff
@@ -1,208 +1,206 @@
...
+    }
+  }
 
-    onToggleInlineTranscript: function(e) {
-      if (e && e.preventDefault) e.preventDefault();
+  onSkipToTranscript() {
+    // need slight delay before focussing button to make it work when JAWS is running
+    // see https://github.com/adaptlearning/adapt_framework/issues/2427
+    _.delay(() => {
+      this.$('.youtube__transcript-btn').a11y_focus();
```
Comment (tomgreenfield) [inline js/adapt-youtube.js]:
```suggestion
      Adapt.a11y.focusFirst(this.$('.youtube__transcript-btn'), { defer: true });
```

File: js/adapt-youtube.js [hunk 5]
```diff
@@ -1,208 +1,206 @@
...
+    const $transcriptBodyContainer = this.$('.youtube__transcript-body-inline');
+    const $button = this.$('.youtube__transcript-btn-inline');
+    const $buttonText = $button.find('.youtube__transcript-btn-text');
 
-        $button.attr('aria-expanded', false);
-        $buttonText.html(this.model.get('_transcript').inlineTranscriptButton);
+    if ($transcriptBodyContainer.hasClass('inline-transcript-open')) {
+      $transcriptBodyContainer.stop(true, true).slideUp(() => {
+        $(window).resize();
+      }).removeClass('inline-transcript-open');
```
Comment (tomgreenfield) [inline js/adapt-youtube.js]: Suggest refactoring here to remove duplication with something like:
```js
const config = this.model.get('_transcript');
const shouldOpen = !$transcriptBodyContainer.hasClass('inline-transcript-open');
const buttonText = shouldOpen ?
  config.inlineTranscriptCloseButton :
  config.inlineTranscriptButton

$transcriptBodyContainer
  .stop(true).slideToggle(() => $(window).resize())
  .toggleClass('inline-transcript-open', shouldOpen);
$button.attr('aria-expanded', shouldOpen);
$buttonText.html(buttonText);

if (!shouldOpen || config._setCompletionOnView === false) return;
this.setCompletionStatus();
```

File: js/adapt-youtube.js [hunk 6]
```diff
@@ -1,208 +1,206 @@
...
-  var YouTubeView = ComponentView.extend({
-    player: null,
-
-    events: {
+import Adapt from 'core/js/adapt';
+import ComponentView from 'core/js/views/componentView';
+import ComponentModel from 'core/js/models/componentModel';
+
+class YouTubeView extends ComponentView {
+
```
Comment (tomgreenfield) [inline js/adapt-youtube.js]: ESLint: unnecessary whitespace.
```suggestion

```

File: js/adapt-youtube.js [hunk 7]
```diff
@@ -1,208 +1,206 @@
...
+    };
+  }
+
+  initialize() {
+    super.initialize();
+
+    _.bindAll(this, 'onPlayerStateChange', 'onPlayerReady', 'onInview');
+
+    this.player = null;
+    this.debouncedTriggerGlobalEvent =_.debounce(this.triggerGlobalEvent.bind(this), 1000);
```
Comment (tomgreenfield) [inline js/adapt-youtube.js]: ESLint: space needed.

```suggestion
    this.debouncedTriggerGlobalEvent = _.debounce(this.triggerGlobalEvent.bind(this), 1000);
```

File: js/adapt-youtube.js [hunk 8]
```diff
@@ -1,208 +1,206 @@
...
+    // default aspect ratio to 16:9 if not specified
+    const aspectRatio = this.model.get('_media')._aspectRatio 
+      ? parseFloat(this.model.get('_media')._aspectRatio)
+      : 1.778;
+    if (isNaN(aspectRatio)) return;
+    $iframe.height(widgetWidth / aspectRatio);
+  }
+
+  postRender() {
+    //FOR HTML/HBS Paramenters: https://developers.google.com/youtube/player_parameters
```
Comment (tomgreenfield) [inline js/adapt-youtube.js]: ESLint: need space.

```suggestion
    // for HTML/HBS parameters: https://developers.google.com/youtube/player_parameters
```

File: js/adapt-youtube.js [hunk 9]
```diff
@@ -1,208 +1,206 @@
...
-    },
-
-    postRender: function() {
-      //FOR HTML/HBS Paramenters: https://developers.google.com/youtube/player_parameters
-      if (Adapt.youTubeIframeAPIReady === true) {
-        this.onYouTubeIframeAPIReady();
-      } else {
-        Adapt.once('youTubeIframeAPIReady', this.onYouTubeIframeAPIReady, this);
+  onYouTubeIframeAPIReady() {
+    this.player = new YT.Player(this.$('iframe').get(0), {
```
Comment (tomgreenfield) [inline js/adapt-youtube.js]: ESLint: undefined global.
```suggestion
    this.player = new window.YT.Player(this.$('iframe').get(0), {
```
See also other instances in `onPlayerStateChange()`.

File: js/adapt-youtube.js [hunk 10]
```diff
@@ -1,208 +1,206 @@
...
+  /**
+  * this seems to have issues in Chrome if the user is logged into YouTube (possibly any Google account) - the API just doesn't broadcast the events
+  * but instead throws the error:
+  * Failed to execute 'postMessage' on 'DOMWindow': The target origin provided ('https://www.youtube.com') does not match the recipient window's origin ('http://www.youtube.com').
+  * This is documented here:
+  *   https://code.google.com/p/gdata-issues/issues/detail?id=5788
+  * but I haven't managed to get any of the workarounds to work... :-(
+  */
+  onPlayerStateChange(e) {
+    switch(e.data) {
```
Comment (tomgreenfield) [inline js/adapt-youtube.js]: ESLint: needs space.
```suggestion
    switch (e.data) {
````

File: js/adapt-youtube.js [hunk 11]
```diff
@@ -1,208 +1,206 @@
...
-      // need slight delay before focussing button to make it work when JAWS is running
-      // see https://github.com/adaptlearning/adapt_framework/issues/2427
-      _.delay(function() {
-        this.$('.youtube__transcript-btn').a11y_focus();
-      }.bind(this), 250);
-    },
+        if (this.model.get('_setCompletionOn') && this.model.get('_setCompletionOn') === 'ended') {
+          this.setCompletionStatus();
+        }
+      break;
```
Comment (tomgreenfield) [inline js/adapt-youtube.js]: ESLint: wrong indentation.
```suggestion
        break;
```
See also the other `break` statements.

File: js/adapt-youtube.js (hunk 5 above)
Comment (tomgreenfield) [inline js/adapt-youtube.js:167]: Suggest refactor to remove duplication with something like:
```js
const config = this.model.get('_transcript');
const shouldOpen = !$transcriptBodyContainer.hasClass('inline-transcript-open');
const buttonText = shouldOpen ?
  config.inlineTranscriptCloseButton :
  config.inlineTranscriptButton;

$transcriptBodyContainer
  .stop(true).slideToggle(() => $(window).resize())
  .toggleClass('inline-transcript-open', shouldOpen);
$button.attr('aria-expanded', shouldOpen);
$buttonText.html(buttonText);

if (!shouldOpen || config._setCompletionOnView === false) return;
this.setCompletionStatus();
```

File: js/adapt-youtube.js [hunk 12]
```diff
@@ -1,208 +1,206 @@
...
-    view: YouTubeView
-  });
+}
 
+Adapt.register('youtube', {
+  model: ComponentModel.extend({}),
+  view: YouTubeView
 });
+
+export default YouTubeView;
```
Comment (tomgreenfield) [inline js/adapt-youtube.js]:
```suggestion
export default Adapt.register('youtube', {
  model: ComponentModel.extend({}),
  view: YouTubeView
});
```

File: js/adapt-youtube.js (hunk 1 above)
Comment (tomgreenfield) [inline js/adapt-youtube.js:39]:
```suggestion
      'device:resize device:changed': this.setIFrameSize,
```

File: js/adapt-youtube.js [hunk 13]
```diff
@@ -1,208 +1,194 @@
...
         break;
-        case YT.PlayerState.ENDED:
-          this.triggerGlobalEvent('ended');
+      case window.YT.PlayerState.ENDED:
+        this.triggerGlobalEvent('ended');
 
-          if (this.model.get('_setCompletionOn') && this.model.get('_setCompletionOn') === 'ended') {
-            this.setCompletionStatus();
-          }
+        if (this.model.get('_setCompletionOn') && this.model.get('_setCompletionOn') === 'ended') {
```
Comment (tomgreenfield) [inline js/adapt-youtube.js]:
```suggestion
        if (this.model.get('_setCompletionOn') === 'ended') {
```
See also line 126.

**Review by tomgreenfield — APPROVED** [review]

**Review by oliverfoster — APPROVED** [review]
